src/codegen/compile.ml

```diff
@@ -4164,7 +4164,81 @@ module RTS_Exports = struct
 
 end (* RTS_Exports *)
 
-module Serialization = struct
+module type Stream = sig
+  (*               env    ref    code *)
+  val write_byte : E.t -> G.t -> G.t -> G.t
+  val write_word_leb : E.t -> G.t -> G.t -> G.t
+  val write_word_32 : E.t -> G.t -> G.t -> G.t
+  val write_blob : E.t -> G.t -> G.t -> G.t
+  val write_text : E.t -> G.t -> G.t -> G.t
+  val write_unsigned : E.t -> G.t -> G.t -> G.t
+  val write_signed : E.t -> G.t -> G.t -> G.t
+
+  (* opportunity to flush or update, stream token is on stack *)
+  val checkpoint : E.t -> G.t -> G.t
```

Comment:
I don't understand this. Do we need it?

Reply:
It is only used in one place (`write`). It calls `serialize_go` in turn, which is a function, so it cannot communicate via locals, only stack. Hence for a `BumpStream` the pointer in the caller needs to be updated. For a `Rust` stream (where `ref` is a handle) no update is needed and `checkpoint` can be `nop`.

Reply:
Ah, in other words, this is to take the steam, put it's (local) data on the stack and pass to a functions or return from one.

If the pointer would be in memory we wouldn't need this, right?

Reply:
Ok, I understand now.

I wonder if an API consisting of `get_local_state` and `set_local_state` would be easier to understand. Although it's kinda silly if the first one always just returns the token. Ah: but if that token was or were abstract, it _would_ make sense.

Reply:
We can always change the stream API as we see fit. The next PR will implement a Rust buffered stream and that may give us hints how this could be improved.